docs/IsolationNetworkPolicies.md

```diff
@@ -0,0 +1,121 @@
+# Isolation based on Kubernetes Network Policies
+
+## Introduction
+
+As described in the [Fybrik Network topologies and Requirements](FybrikNetworkTopologiesAndRequirements.md)  we have to 
+restrict network traffic of the data plane components to only allowed connections. One of the methods that can be used is 
+Kubernetes Network Policies. This document explains the feature. 
+
+## Network Policies
+
+Fybrik is a Kubernetes application, therefore, the simplest isolation method can be based on 
+[Kubernetes Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) (NP).
+NP allow to control traffic flow at the IP address or port level (OSI layer 3 or 4). NP can restrict incoming (ingress) 
+and/or outgoing (egress) communications. 
+The restrictions are based on combinations of Pods, Namespaces labels and IP blocks.
+
+### Advantages:
+- does not require any changes in module implementations, can be implemented in the Fybrik control plane only.
+  - A single possible module definition extension is to declare in the module manifest external services that a module 
+    needs to connect, see below. 
+- can be combined with other isolation methods (future work).
+
+### Disadvantages:
+- a single Kubernetes cluster oriented, cross clusters restrictions can be based on IP blocks only, which significantly 
+complicates the configuration. Therefore, later we will provide other isolation methods as an extension to NP.
+- NP are implemented by the network plugin. Creating a NetworkPolicy resource without a controller - a Container Network 
+Interface (CNI) that implements it will have no effect. For example, `Kind` deployments which by default use "kindnetd" 
+do not support NP. In order to support NP, installation of another CNI, e.g. [Calico](https://github.com/projectcalico/calico)
+is required. Here is the [instructions](https://alexbrand.dev/post/creating-a-kind-cluster-with-calico-networking/). 
+On the other hand, all K8s clusters in production deployments use CNIs which support NP.
+
+## Implementation
+
+The `from`/`to` element of `ingress`/`egress`  
+[NetworkPolicySpec](https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/network-policy-v1/#NetworkPolicySpec) 
+allows to restrict incoming/outgoing network traffic based on `podSelectors` (labels based), and/or 
+`namespaceSelectors` (labels based) and/or `ipBlocks`. The last parameter can be used to restrict cross cluster connections 
+
+We suggest, after deployment a data plane module, to create an instance of a network policy that will restrict the 
+module connectivity.
```

Comment:
> Who is responsible for the network policy creation - Blueprint controller?

Yes, there is an explanation about it:
"BlueprintController install and uninstall Fybrik modules. We suggest extending this functionality by creating an instance of NP for each installing module, and delete it when the module is uninstalled." 

> The module status will be marked as ready only after the network policy is deployed, correct?
Actually, we can deploy NP **before** module deployment, so the above statement will be correct. (updated the explanation)

> Does the deployment take effect immediately? 
Yes, I've never seen any comments about NP delay, however they do not influence on the existing connections, so deployment NP before the relevant module is better. 

> Should some resource status be checked?
Usually we don't have to. We can validate that NP ingress/egress definitions match the incoming/outgoing trafic. 
If somebody insist he can run netcat([nc](https://linux.die.net/man/1/nc)) on a pod. It can be part of our integration tests. 
NetworkPolicies are an optional CNI feature, and if a CNI doesn't support NP (e.g. kind by default), no error will be sent, but all connections will be allowed. 

> What happens in case of a failure (on the network plugin side) - will the error be propagated to the user?
I'm not sure that I understand your question, do you mean error in the CNI implementation? It depends on CNI, some of them can create Kubernetes event, most of them write an error/warning into **their** logs.

@zivnevo told me that NP-gurard could help to debug a system.